zerver/lib/actions.py

```diff
@@ -5036,6 +5036,52 @@ def do_mark_stream_messages_as_read(
     return count
 
 
+def do_mark_user_messages_as_read(
+    user_profile: UserProfile,
+    muted_user_profile: UserProfile,
+) -> int:
+    msgs = UserMessage.objects.filter(
+        user_profile=user_profile,
+    )
+
+    msgs = msgs.filter(message__sender=muted_user_profile)
+
+    msgs = msgs.extra(
+        where=[UserMessage.where_unread()],
+    )
+
+    message_ids = list(msgs.values_list("message__id", flat=True))
+
+    count = msgs.update(
+        flags=F("flags").bitor(UserMessage.flags.read),
+    )
+
+    event = dict(
+        type="update_message_flags",
+        op="add",
+        operation="add",
+        flag="read",
+        messages=message_ids,
+        all=False,
+    )
+    event_time = timezone_now()
```

Comment:
It would be a nice preparatory cleanup to create a `dataclass` for the `update_message_flags` event type so we stop duplicating these dictionaries, but would also be a fine followup, so feel free to just copy this idea onto your TODO list and focus on the rest.